### PR TITLE
164798016 Feature password reset 

### PR DESCRIPTION
**Pull request by vic3king**
#### What does this PR do?
This PR adds Password reset functionality via email
#### Description of Task to be completed?
- create routes requires
- create controllers for password reset
- validate input fields
- install npm packages
- add email notification
#### How should this be manually tested?
check out ft-password-reset-164798016
npm run start:dev
make a post request to api/v1/auth/reset with the email address on the body
retrieve the token from email
make a post request with new password to api/v1/auth/new_password
#### Any background context you want to provide?
I will raise another PR for the tests and documentation
#### What are the relevant pivotal tracker stories?
[164798016](https://www.pivotaltracker.com/n/projects/2318856/stories/164798016)

### CODE REVIEW

File: server/controllers/resetpassword.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import Authenticate from '../helpers/auth';
+import model from '../models';
+
+const { User } = model;
+
+const { hashPassword } = Authenticate;
+
+const ResetPasswordController = {
+  async updatePassword(req, res) {
+    const { userObj } = req.user;
```
Comment (rotimi-babalola) [inline server/controllers/resetpassword.js]: But have a different function to sign the user's email to avoid this naming confusion

**Review by rotimi-babalola — CHANGES_REQUESTED** [review]
Looks good just work on the feedback and wrap `await` statements in `try/catch` blocks

**Comment by rotimi-babalola** [comment]
18 files changed? Could this be smaller in the future? 🤔

**Comment by vic3king** [comment]
> 18 files changed? Could this be smaller in the future? 🤔

Definitely. I added tests and it got bloated. I'll make sure they are smaller moving forward.

File: server/middlewares/resetpassword.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+  },
+
+  async validateToken(req, res, next) {
+    try {
+      req.data = decode(req.params.token);
+    } catch (error) {
+      if (error) {
+        return res.status(406).json({
+          status: 406,
```
Comment (rotimi-babalola) [inline server/middlewares/resetpassword.js:69]: Why are we sending `406` here? @vic3king

Comment (vic3king) [reply server/middlewares/resetpassword.js:69]: @senor I used that error to catch a situation for any reason where the token being decoded fails.

**Comment by rotimi-babalola** [comment]
There are merge conflicts 🙂 @vic3king

**Comment by vic3king** [comment]
> There are merge conflicts 🙂 @vic3king

@rotimi-babalola Fixed 🙂

**Pull request merged**